File: web/src/api/forecasts.ts
import { useQuery } from "@tanstack/react-query";

import type { OutcomeBucket } from "../lib/charts";
import { apiFetch } from "./client";
import { metricsParams, type MetricsFilters, type MetricsScope } from "./metrics";

export type { OutcomeBucket };

export interface CompletionForecast {
  trials: number;
  p50_date: string;
  p75_date: string;
  p85_date: string;
  p95_date: string;
  outcomes: OutcomeBucket[];
}

export interface Forecast {
  window_start: string;
  window_end: string;
  remaining: number;
  completion: CompletionForecast | null;
  confidence: number | null;
}

export interface ForecastOptions {
  /** Only the item filters are forwarded — see the comment below. */
  filters?: MetricsFilters;
  /** ISO YYYY-MM-DD; asks the API for the odds of finishing by this date. */
  targetDate?: string;
  /** Scenario override for the backlog size; omitted means "the real count". */
  remaining?: number;
}

export function useForecast(scope: MetricsScope, options: ForecastOptions = {}) {
  const { filters, targetDate, remaining } = options;
  // The page's period is deliberately NOT forwarded: window_end doubles as the
  // forecast origin, so a custom past range would turn "P85 finish" into a
  // backtest date instead of a prediction. Passing only the item-filter subset
  // to metricsParams makes that structural rather than a rule to remember.
  const params = metricsParams(scope, {
    types: filters?.types,
    excludeStates: filters?.excludeStates,
  });
  const search = new URLSearchParams(params ?? "");
  if (targetDate) search.set("target_date", targetDate);
  if (remaining !== undefined) search.set("remaining", String(remaining));
  return useQuery({
    queryKey: ["forecasts", scope, filters?.types, filters?.excludeStates, targetDate ?? null, remaining ?? null],
    enabled: params !== null,
    queryFn: () => apiFetch<Forecast>(`/api/forecasts?${search.toString()}`),
  });
}

File: app/api/forecasts.py
from datetime import date, datetime, timedelta

from fastapi import APIRouter, Query

from app.api.deps import ForecastServiceDep
from app.api.schemas import CompletionForecastRead, ForecastRead, OutcomeBucketRead
from app.api.scope import ScopeDep
from app.domain.forecasting.monte_carlo import CompletionForecast

router = APIRouter(prefix="/api/forecasts", tags=["forecasts"])


def _completion_read(
    completion: CompletionForecast | None, origin: datetime
) -> CompletionForecastRead | None:
    if completion is None:
        return None
    return CompletionForecastRead(
        trials=completion.trials,
        p50_date=origin + timedelta(days=completion.p50_days),
        p75_date=origin + timedelta(days=completion.p75_days),
        p85_date=origin + timedelta(days=completion.p85_days),
        p95_date=origin + timedelta(days=completion.p95_days),
        outcomes=[
            OutcomeBucketRead(days=o.days, trials=o.trials) for o in completion.outcomes
        ],
    )


@router.get("", response_model=ForecastRead)
async def get_forecast(
    service: ForecastServiceDep,
    scope: ScopeDep,
    window_days: int = Query(default=90, ge=7, le=365),
    remaining: int | None = Query(default=None, ge=0),
    target_date: date | None = None,
) -> ForecastRead:
    forecast = await service.get_forecast(
        team_id=scope.team_id,
        project_id=scope.project_id,
        window_days=window_days,
        remaining=remaining,
        target_date=target_date,
    )
    return ForecastRead(
        window_start=forecast.window_start,
        window_end=forecast.window_end,
        remaining=forecast.remaining,
        completion=_completion_read(forecast.completion, forecast.window_end),
        confidence=forecast.confidence,
    )
